effort_controllers/test/simple_bot_hw.cpp

```diff
@@ -0,0 +1,120 @@
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//////////////////////////////////////////////////////////////////////////////
+
+// ROS
+#include <chrono>
+#include <thread>
+#include <ros/ros.h>
+#include <std_msgs/Float64.h>
+#include <rosgraph_msgs/Clock.h>
+
+// ros_control
+#include <controller_manager/controller_manager.h>
+#include <hardware_interface/joint_command_interface.h>
+#include <hardware_interface/joint_state_interface.h>
+#include <hardware_interface/robot_hw.h>
+#include <realtime_tools/realtime_buffer.h>
+
+class SimpleBot : public hardware_interface::RobotHW
+{
+public:
+  static constexpr double period = 0.01;
+
+  SimpleBot()
+  : pos_(0)
+  , vel_(0)
+  , eff_(0)
+  , max_eff_(10)
+  {
+    // joint state interface
+    hardware_interface::JointStateHandle state_handle("joint", &pos_, &vel_, &eff_);
+    state_interface_.registerHandle(state_handle);
+    registerInterface(&state_interface_);
+    // effort interface
+    hardware_interface::JointHandle effort_handle(state_interface_.getHandle("joint"), &cmd_);
+    effort_interface_.registerHandle(effort_handle);
+    registerInterface(&effort_interface_);
+  }
+
+  void read() {
+    // update current state
+    pos_ += period * vel_ + 0.5*std::pow(period, 2.) * eff_;
+    vel_ += period * eff_;
+  }
+
+  void write() {
+    // store current command
+    eff_ = cmd_;
+    if(std::fabs(eff_) > max_eff_) {
+      ROS_WARN("Large effort command received: %f; cutting down to % f", eff_, max_eff_);
+      eff_ = eff_ > 0 ? max_eff_ : -max_eff_;
+    }
+  }
+
+private:
+  hardware_interface::JointStateInterface state_interface_;
+  hardware_interface::EffortJointInterface effort_interface_;
+  double cmd_, pos_, vel_, eff_;
+  double max_eff_;
+};
+
+
+int main(int argc, char **argv)
+{
+  ros::init(argc, argv, "simple_bot_hw");
+  ros::NodeHandle nh;
+  SimpleBot bot;
+  ros::Publisher clock_publisher = nh.advertise<rosgraph_msgs::Clock>("/clock", 1);
+  nh.setParam("/use_sim_time", true);
+  controller_manager::ControllerManager cm(&bot, nh);
+  ros::AsyncSpinner spinner(2);
+  spinner.start();
+
+  ros::Rate rate(1./bot.period);
+  ros::Duration period(bot.period);
+  double sim_time = 0;
+
+  // main "simulation"
+  while(ros::ok()) {
+    auto begin = std::chrono::system_clock::now();
+    bot.read();
+    cm.update(ros::Time(sim_time), period);
+    bot.write();
+    auto end = std::chrono::system_clock::now();
+    double elapsed = std::chrono::duration_cast<std::chrono::duration<double> >((end - begin)).count();
+    if(bot.period > elapsed)
+      std::this_thread::sleep_for(std::chrono::duration<double>(bot.period - elapsed));
+    rosgraph_msgs::Clock clock;
+    clock.clock = ros::Time(sim_time);
+    clock_publisher.publish(clock);
+    sim_time += bot.period;
```

Comment:
To be honest, I decided to put it because in the first versions of the test I was getting no errors on my PC but checks were failing during CI. I imagined that one issue was related to the virtual machine: perhaps it had limited processing power and thus I wanted to "slow down" things accordingly at least in the RobotHW. I saw that other packages were using the simulated clock and I thus used the same procedure. Now that I found I forgot to initialize the `cmd_` variable, I think that was the main source of issues 😅 But I think it might still be useful, in case the VM slows down...